### ISSUE TITLE
Please migrate away from setup.py

### ISSUE THREAD

**Issue by baldurmen** [alan-turing-institute/CleverCSV#76]
Hello!

`setup.py` has been deprecated for a while now (although support hasn't yet bet removed). It would be nice if this project moved away from it before something actually breaks :)

If you want to stick with `setuptools`, newer versions do support building via a PEP-617-style `pyproject.toml` file:

* https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html
* https://peps.python.org/pep-0621

You can see an example of a migration to `pyproject.toml` with setuptools that I did here: https://github.com/jaseg/python-mpv/pull/241

Cheers,

**Comment by SnoopJ** [comment]
Point of clarification: `setup.py` is not itself deprecated (unless I am misinformed, in which case I would appreciate a pointer to the deprecation notice), but [_running_ `setup.py` directly](https://github.com/alan-turing-institute/CleverCSV/blob/3bdae29f6125682639d5253d62ab782a87349513/Makefile#L30-L31) is [deprecated](https://setuptools.pypa.io/en/latest/deprecated/commands.html). I'm not familiar enough with this project to know if the executable portions of `setup.py` could be subsumed into an entirely-declarative way (`setup.cfg` or `pyproject.toml`) but it would be a good improvement to remove the direct execution of the file from the build tooling.

**Comment by sanmai-NL** [comment]
@baldurmen It seems to have been fixed. Can this be closed?

**Comment by baldurmen** [comment]
Hi,

I guess this issue could be closed, yes. I would prefer if it wasn't though, as I don't think the current fix is what I intended :)

@SnoopJ is right in correcting me by saying `setup.py` isn't deprecated itself (only running it as CLI argument to build is). They're also right when they say you can now build this project using `python3.11 -m build` (I've just tested this while updating the Debian package to 0.8.0).

Most of the logic stays in the `setup.py` file though, and is in-house code. I think the best practice would be to migrate most (if not all) of that logic in the `pyproject.toml` file :)

I had a look and it doesn't seem like it would be a lot of work?

**Comment by sanmai-NL** [comment]
Yes you're right, the `pyproject.toml` is there but the `setup.py` lingers.

**Comment by GjjvdBurg** [comment]
Thanks both for the discussion! I looked into this a bit but haven't fully completed the transition. I believe from what I've seen so far that the `setup.py` file is still needed for some build functionality (such as [the C extensions](https://setuptools.pypa.io/en/latest/userguide/ext_modules.html)), so my thinking is that in order to resolve this ticket it would be best to have that be the _only_ things that are defined in the `setup.py` file. Until then I'll leave this ticket open.

**Comment by sanmai-NL** [comment]
You can use PDM: https://pdm-backend.fming.dev/build_config/#run-setuptools
